src/impl/mpispace/Kokkos_MPISpace_Ops.hpp

```diff
@@ -71,108 +66,155 @@ KOKKOS_REMOTESPACES_G(short, MPI_SHORT)
-  }
-
-KOKKOS_REMOTESPACES_ATOMIC_SET(int)
-KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned int)
-KOKKOS_REMOTESPACES_ATOMIC_SET(long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(long long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned long long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(float)
-KOKKOS_REMOTESPACES_ATOMIC_SET(double)
-
+#define KOKKOS_REMOTESPACES_ATOMIC_SET(type, mpi_type)                        \
+  static KOKKOS_INLINE_FUNCTION void mpi_type_atomic_set(                     \
+      const type &val, int offset, int pe, const MPI_Win &win) {              \
+    MPI_Accumulate(&val, 1, mpi_type, pe,                                     \
+                   sizeof(SharedAllocationHeader) + offset * sizeof(type), 1, \
+                   mpi_type, MPI_REPLACE, win);                               \
+    MPI_Win_flush(pe, win);                                                   \
+  }
+
+KOKKOS_REMOTESPACES_ATOMIC_SET(char, MPI_SIGNED_CHAR)
+KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned char, MPI_UNSIGNED_CHAR)
+KOKKOS_REMOTESPACES_ATOMIC_SET(short, MPI_SHORT)
+KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned short, MPI_UNSIGNED_SHORT)
+KOKKOS_REMOTESPACES_ATOMIC_SET(int, MPI_INT)
+KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned int, MPI_UNSIGNED)
+KOKKOS_REMOTESPACES_ATOMIC_SET(long, MPI_LONG)
+KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned long, MPI_UNSIGNED_LONG)
+KOKKOS_REMOTESPACES_ATOMIC_SET(long long, MPI_LONG_LONG)
+KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned long long, MPI_UNSIGNED_LONG_LONG)
+KOKKOS_REMOTESPACES_ATOMIC_SET(float, MPI_FLOAT)
+KOKKOS_REMOTESPACES_ATOMIC_SET(double, MPI_DOUBLE)
 #undef KOKKOS_REMOTESPACES_ATOMIC_SET
 
-#define KOKKOS_REMOTESPACES_ATOMIC_FETCH(type)                        \
-  static KOKKOS_INLINE_FUNCTION type mpi_type_atomic_fetch(type *ptr, \
-                                                           int pe) {  \
-    /* TBD */                                                         \
-    return 0;                                                         \
-  }
-
-KOKKOS_REMOTESPACES_ATOMIC_FETCH(int)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH(unsigned int)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH(long)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH(unsigned long)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH(long long)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH(unsigned long long)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH(float)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH(double)
-
+#define KOKKOS_REMOTESPACES_ATOMIC_FETCH(type, mpi_type)                     \
+  static KOKKOS_INLINE_FUNCTION type mpi_type_atomic_fetch(                  \
+      type val, int offset, int pe, const MPI_Win &win) {                    \
+    type ret;                                                                \
+    MPI_Fetch_and_op(&val, &ret, mpi_type, pe,                               \
+                     sizeof(SharedAllocationHeader) + offset * sizeof(type), \
+                     MPI_OP_NULL, win);                                      \
```

Comment:
This should be `MPI_NO_OP`, not `MPI_OP_NULL`. Besides that, why do you use an atomic operation for `operator-` etc?

Reply:
Correcting to MPI_NO_OP. We implemented those operators with an atomic fetch in order to preserve correct ordering (acquire memory semantics) with other operations.